Index: plugins/file/FilePlugin.cs
===================================================================
--- plugins/file/FilePlugin.cs	(revision 80)
+++ plugins/file/FilePlugin.cs	(working copy)
@@ -3,8 +3,6 @@
 using System.IO;
 using System.Collections;
 using System.Threading;
-using Entagged;
-using Entagged.Audioformats.Exceptions;
 using Nini;
 using DAAP;
 using log4net;
@@ -185,11 +183,9 @@
             if (songHash[file] != null)
                 return;
             
-            AudioFile af;
+            TagLib.File af = TagLib.File.Create (file);
 
-            try {
-                af = new AudioFile (file);
-            } catch (Exception e) {
+            if (af == null) {
                 return;
             }
 
@@ -199,20 +195,20 @@
                 db.AddTrack (song);
             }
 
-            song.Artist = af.Artist;
-            song.Album = af.Album;
-            song.Title = af.Title;
-            song.Duration = af.Duration;
+            song.Artist = (new TagLib.StringList (af.Tag.Artists)).ToString ("; ");
+            song.Album = af.Tag.Album;
+            song.Title = af.Tag.Title;
+            song.Duration = new System.TimeSpan (0, 0, af.AudioProperties.Length);
             song.FileName = file;
             song.Format = Path.GetExtension (file).Substring (1);
-            song.Genre = af.Genre;
+            song.Genre = (new TagLib.StringList (af.Tag.Genres)).ToString ("; ");
 
             FileInfo info = new FileInfo (file);
             song.Size = (int) info.Length;
-            song.TrackCount = af.TrackCount;
-            song.TrackNumber = af.TrackNumber;
-            song.Year = af.Year;
-            song.BitRate = (short) af.Bitrate;
+            song.TrackCount = (int) af.Tag.TrackCount;
+            song.TrackNumber = (int) af.Tag.Track;
+            song.Year = (int) af.Tag.Year;
+            song.BitRate = (short) af.AudioProperties.Bitrate;
 
             songHash[file] = song;
 
Index: plugins/file/Makefile.am
===================================================================
--- plugins/file/Makefile.am	(revision 80)
+++ plugins/file/Makefile.am	(working copy)
@@ -11,7 +11,7 @@
 PLUGIN_DEPS = 						\
 	-r:$(top_builddir)/deps/daap-sharp/daap-sharp.dll	\
 	-r:$(top_srcdir)/deps/Nini.dll			\
-	-r:$(top_srcdir)/deps/entagged-sharp.dll	\
+        $(TAGLIB_SHARP_LIBS)				\
 	-r:$(srcdir)/db4o.dll				\
 	-r:$(top_srcdir)/deps/log4net.dll		\
 	-r:Mono.Posix					\
Index: configure.ac
===================================================================
--- configure.ac	(revision 80)
+++ configure.ac	(working copy)
@@ -47,6 +47,9 @@
 PKG_CHECK_MODULES(GLADE_SHARP, glade-sharp-2.0)
 AC_SUBST(GLADE_SHARP_LIBS)
 
+PKG_CHECK_MODULES(TAGLIB_SHARP, taglib-sharp)
+AC_SUBST(TAGLIB_SHARP_LIBS)
+
 AC_ARG_ENABLE(beagle, AC_HELP_STRING([--disable-beagle], [Do not build the beagle plugin]), enable_beagle="no", enable_beagle="yes")
 
 if test "x$enable_beagle" = "xyes"; then \
